src/java/org/apache/cassandra/index/sai/StorageAttachedIndexBuilder.java

```diff
@@ -0,0 +1,351 @@
+        {
+            SSTableReader sstable = e.getKey();
+            Set<StorageAttachedIndex> indexes = e.getValue();
+
+            Set<StorageAttachedIndex> existing = validateIndexes(indexes, sstable.descriptor);
+            if (existing.isEmpty())
+            {
+                logger.debug(logMessage("{} dropped during index build"), indexes);
+                continue;
+            }
+
+            if (indexSSTable(sstable, existing))
+            {
+                return;
+            }
+        }
+    }
+
+    private String logMessage(String message) {
+        return String.format("[%s.%s.*] %s", metadata.keyspace, metadata.name, message);
+    }
+
+    /**
+     * @return true if index build should be stopped
+     */
+    private boolean indexSSTable(SSTableReader sstable, Set<StorageAttachedIndex> indexes)
+    {
+        logger.debug(logMessage("Starting index build on {}"), sstable.descriptor);
+
+        CountDownLatch perSSTableFileLock = null;
+        StorageAttachedIndexWriter indexWriter = null;
+
+        Ref<? extends SSTableReader> ref = sstable.tryRef();
+        if (ref == null)
+        {
+            logger.warn(logMessage("Couldn't acquire reference to the SSTable {}. It may have been removed."), sstable.descriptor);
+            return false;
+        }
+
+        try (RandomAccessReader dataFile = sstable.openDataReader();
+             LifecycleTransaction txn = LifecycleTransaction.offline(OperationType.INDEX_BUILD, sstable))
+        {
+            perSSTableFileLock = shouldWritePerSSTableFiles(sstable);
+            // If we were unable to get the per-SSTable file lock it means that the
+            // per-SSTable components are already being built, so we only want to
+            // build the per-index components
+            boolean perIndexComponentsOnly = perSSTableFileLock == null;
+            // remove existing per column index files instead of overwriting
+            IndexDescriptor indexDescriptor = IndexDescriptor.create(sstable);
+            indexes.forEach(index -> indexDescriptor.deleteColumnIndex(index.getIndexContext()));
+
+            indexWriter = new StorageAttachedIndexWriter(indexDescriptor, indexes, txn, perIndexComponentsOnly);
+
+            long previousKeyPosition = 0;
+            indexWriter.begin();
+
+            try (KeyIterator keys = new KeyIterator(sstable.descriptor, metadata))
+            {
+                while (keys.hasNext())
+                {
+                    if (isStopRequested())
+                    {
+                        logger.debug(indexDescriptor.logMessage("Index build has been stopped"));
+                        throw new CompactionInterruptedException(getCompactionInfo());
+                    }
+
+                    final DecoratedKey key = keys.next();
+                    final long keyPosition = keys.getKeyPosition();
+
+                    indexWriter.startPartition(key, keyPosition);
```

Comment:
Kind of funny that we don't even need the position anymore. I guess we can remove it later when SASI goes away...

Reply:
It was used pre row-awareness for the key position but we don't use it now. I suppose removing it could be done but it would also make sense to keep it there if a future index implementation needed it.